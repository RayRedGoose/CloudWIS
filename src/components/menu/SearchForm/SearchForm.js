import React, { Component } from 'react'
import './SearchForm.scss'

class SearchForm extends Component {
  constructor(props) {
    super();
    this.state = {
      searchCity: ''
    }
  }

  handleChange = event => {
    this.setState({ searchCity: event.target.value });
  }

  keyPressed = event => {
    if (event.key === "Enter") {
      this.submitSearch(event)
    }
  }

  clearInputs = event => {
    event.preventDefault()
    this.setState({ searchCity: '' });
  }

  submitSearch = event => {
    event.preventDefault()
    this.props.changeCity(this.state.searchCity)
    this.props.toggleActiveState()
    this.props.fetchWeatherData(this.state.searchCity)
    this.clearInputs(event)
  }

  render() {
    return (
      <form className="city-search__form">
        <label htmlFor="city-search__input">Find city</label>
        <input
          type="text"
          id="city-search__input"
          placeholder="Enter city here"
          onChange={this.handleChange}
          value={this.state.searchCity}
          onKeyPress={this.keyPressed}/>
        <div className="button-wrapper">
          <button
            id="clear-button"
            onClick={this.clearInputs}>Clear</button>
          <button
            id="submit-button"
            onClick={this.submitSearch}>Search</button>
        </div>
      </form>
    )
  }
}

export default SearchForm
