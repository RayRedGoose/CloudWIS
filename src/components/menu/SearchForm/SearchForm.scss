.city-search__form {
  display: flex;
  flex-direction: column;
  label {
    font-size: 24px;
    padding-bottom: 5px;
    border-bottom: 2px solid #0D324D;
    margin-right: 80px;
  }
  input {
    border-radius: 10px;
    font-size: 18px;
    padding: 0.4em;
    padding-left: 20px;
    margin: 40px 0 20px;
  }

}

.button-wrapper {
  display: flex;
  justify-content: space-between;
  button {
    cursor: pointer;
    border-radius: 20px;
    font-size: 18px;
    font-weight: 700;
    padding: 0.4em 2em;
    width: fit-content;
    &:last-child {
      background-color: lighten(#0D324D, 10%);
      color: #fff;
    }
    &:first-child {
      background-color: #fff;
      color: #0D324D;
    }
    &:hover {
      background-color: lighten(#0D324D, 15%);
      color: #fff;
    }
    &:active {
      background-color: darken(#0D324D, 5%);
    }
  }
}
