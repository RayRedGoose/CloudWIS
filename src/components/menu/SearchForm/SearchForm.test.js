import React from 'react';
import { shallow } from 'enzyme';
import SearchForm from './SearchForm';

describe('SearchForm', () => {
  let form, mockProps, mockEvent
  const changeCity = jest.fn()
  const fetchWeatherData = jest.fn()
  const toggleActiveState = jest.fn()

  beforeEach(() => {
    mockProps = {
      changeCity: changeCity,
      fetchWeatherData: fetchWeatherData,
      toggleActiveState: toggleActiveState
    }

    form = shallow(<SearchForm { ...mockProps }/>);
  })

  it('should match the snapshot if it\'s not clicked', () => {
    expect(form).toMatchSnapshot();
  })

  it('should update sought-for city when input change', () => {

    expect(form.state('searchCity')).toEqual('');

    form.find('input').simulate(
      'change',
      {target: {value: 'Atlanta'}}
    );

    expect(form.state('searchCity')).toEqual('Atlanta');
  });

  it('should call handleChange when input is changed', () => {
    const spy = jest.spyOn(form.instance(), 'handleChange').mockImplementation(() => {});
    form.instance().forceUpdate();

    form.find('input').simulate(
      'change',
      { target: { value: 'Atlanta'}
    });

    expect(spy).toHaveBeenCalled();
  });

  it('should clear city state when clear button was pushed', () => {
    form.setState({searchCity: 'Atlanta'})

    expect(form.state('searchCity')).toEqual('Atlanta');

    form.find('#clear-button').simulate(
      'click',
      {preventDefault: jest.fn()}
    );

    expect(form.state('searchCity')).toEqual('');
  });

  it('should call submitSearch when search was activated by enter', () => {
    const spy = jest.spyOn(form.instance(), 'submitSearch').mockImplementation(() => {});
    form.instance().forceUpdate();

    mockEvent = {
      target: document.querySelector('input'),
      key: 'Enter'
    }

    form.find('input').simulate('keypress', mockEvent);

    expect(spy).toHaveBeenCalled();
  });

  describe('submitSearch method', () => {
    beforeEach(() => {
      mockEvent = {
        preventDefault: jest.fn()
      }
    });

    it('should call changeCity prop when search is submitted', () => {
      form.find('#submit-button').simulate('click', mockEvent);

      expect(changeCity).toHaveBeenCalled();
    });

    it('should call toggleActiveState prop when search is submitted', () => {
      form.find('#submit-button').simulate('click', mockEvent);

      expect(toggleActiveState).toHaveBeenCalled();
    });

    it('should call fetchWeatherData prop when search is submitted', () => {
      form.find('#submit-button').simulate('click', mockEvent);

      expect(fetchWeatherData).toHaveBeenCalled();
    })

    it('should call clearInputs when search was submitted', () => {
      const spy = jest.spyOn(form.instance(), 'clearInputs').mockImplementation(() => {});
      form.instance().forceUpdate();

      form.find('#submit-button').simulate('click', mockEvent);

      expect(spy).toHaveBeenCalled();
    })
  })
})
