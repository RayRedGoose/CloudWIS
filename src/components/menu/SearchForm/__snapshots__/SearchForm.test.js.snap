// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should match the snapshot if it's not clicked 1`] = `
<form
  className="city-search__form"
>
  <label
    htmlFor="city-search__input"
  >
    Find city
  </label>
  <input
    id="city-search__input"
    onChange={[Function]}
    onKeyPress={[Function]}
    placeholder="Enter city here"
    type="text"
    value=""
  />
  <div
    className="button-wrapper"
  >
    <button
      id="clear-button"
      onClick={[Function]}
    >
      Clear
    </button>
    <button
      id="submit-button"
      onClick={[Function]}
    >
      Search
    </button>
  </div>
</form>
`;
